chain-impl-mockchain/src/transaction/mod.rs

```diff
@@ -21,6 +21,62 @@ pub struct AuthenticatedTransaction<OutAddress, Extra> {
     pub witnesses: Vec<Witness>,
 }
 
+custom_error! {
+    #[derive(Clone, PartialEq, Eq)]
+    pub VerifyError
+        TooManyInputs {expected: usize, actual: usize }
+            = "too many inputs, expected maximum of {expected}, but received {actual}",
+        TooManyOutputs {expected: usize, actual: usize }
+            = "too many outputs, expected maximum of {expected}, but received {actual}",
+        TooManyWitnesses {expected: usize, actual: usize }
+            = "too many witnesses, expected maximum of {expected}, but received {actual}",
+        NumberOfSignaturesInvalid { expected: usize, actual: usize }
+            = "invalid number of signatures, expected {expected}, but received {actual}",
+}
+
+pub struct VerificationLimits {
+    pub max_inputs_count: usize,
+    pub max_outputs_count: usize,
+    pub max_witnesses_count: usize,
+}
+
+impl<OutAddress, Extra> AuthenticatedTransaction<OutAddress, Extra> {
+    pub fn verify_well_formed(&self, limits: &VerificationLimits) -> Result<(), VerifyError> {
+        let inputs = &self.transaction.inputs;
+        if inputs.len() > limits.max_inputs_count {
+            Err(VerifyError::TooManyInputs {
+                expected: limits.max_inputs_count,
+                actual: inputs.len(),
+            })?
+        }
+
+        let outputs = &self.transaction.outputs;
+        if outputs.len() > limits.max_outputs_count {
+            Err(VerifyError::TooManyOutputs {
+                expected: limits.max_outputs_count,
+                actual: outputs.len(),
+            })?
+        }
+
+        let witnesses = &self.witnesses;
+        if witnesses.len() > limits.max_witnesses_count {
+            Err(VerifyError::TooManyWitnesses {
+                expected: limits.max_witnesses_count,
+                actual: witnesses.len(),
+            })?
+        }
+
+        if inputs.len() != witnesses.len() {
+            Err(VerifyError::NumberOfSignaturesInvalid {
+                expected: inputs.len(),
+                actual: witnesses.len(),
+            })?
+        }
+
+        Ok(())
+    }
+}
```

Comment:
all of this should be in `ledger/check.rs.` Vincent has put a macro to help with the check making the test more consise (can be extended obviously).

Reply:
That's the first place I've tried, but it felt wrong:
1. `check.rs` is private. It looks like a bag of internal helpers and probably shouldn't be made public.
2. This is strictly `AuthenticatedTransaction` logic. It should be as close to its definition as possible, making it method is also convenient and easy to find for users.

Reply:
1. make it public.
2. we are the user. Making it in one module is convenient too.

Reply:
I've made `check.rs` public. `verify_well_formed` is moved to `check.rs` completely. Because it's in the same crate, it can still be a method of `AuthenticatedTransaction`.